File: src/java/PhantasyIslandJavaRemoteControl/radio/TypeDefView.java
package PhantasyIslandJavaRemoteControl.radio;

import java.util.HashMap;
import java.util.Map;

/**
 * 修改 mesh 可视化材质的指令。
 */
public class TypeDefView {

    public static class ViewMaterialChangeCommand {
        public String materialType = null;
        public String color = null;
        public Double opacity = null;
        public Boolean transparent = null;
        public Boolean visible = null;
        public Boolean wireframe = null;
        public Integer side = null;
        public Boolean depthTest = null;
        public Boolean depthWrite = null;
        public Double alphaTest = null;
        public Integer blending = null;
        public Boolean vertexColors = null;
        public Boolean fog = null;
        public String emissive = null;
        public Double emissiveIntensity = null;
        public Double metalness = null;
        public Double roughness = null;
        public Double envMapIntensity = null;
        public Boolean flatShading = null;
        public Double shininess = null;
        public String specular = null;
        public Double clearcoat = null;
        public Double clearcoatRoughness = null;
        public Double transmission = null;
        public Double ior = null;
        public Double thickness = null;
        public Double sheen = null;
        public Double sheenRoughness = null;
        public String sheenColor = null;
        public String attenuationColor = null;
        public Double attenuationDistance = null;
        public Double iridescence = null;
        public Double iridescenceIOR = null;
        public double[] iridescenceThicknessRange = null;
        public Double specularIntensity = null;
        public String specularColor = null;
        public Double reflectivity = null;
        public Double dispersion = null;
        public Double anisotropy = null;
        public Double anisotropyRotation = null;
        public String mapUrl = null;
        public String normalMapUrl = null;
        public String roughnessMapUrl = null;
        public String metalnessMapUrl = null;
        public String emissiveMapUrl = null;
        public String aoMapUrl = null;
        public String alphaMapUrl = null;
        public String bumpMapUrl = null;
        public String displacementMapUrl = null;
        public double[] normalScale = null;
        public Double bumpScale = null;
        public Double displacementScale = null;
        public Double displacementBias = null;
        public Double aoMapIntensity = null;
        public double[] mapRepeat = null;
        public double[] mapOffset = null;
        public Double mapRotation = null;

        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            if (materialType != null) map.put("materialType", materialType);
            if (color != null) map.put("color", color);
            if (opacity != null) map.put("opacity", opacity);
            if (transparent != null) map.put("transparent", transparent);
            if (visible != null) map.put("visible", visible);
            if (wireframe != null) map.put("wireframe", wireframe);
            if (side != null) map.put("side", side);
            if (depthTest != null) map.put("depthTest", depthTest);
            if (depthWrite != null) map.put("depthWrite", depthWrite);
            if (alphaTest != null) map.put("alphaTest", alphaTest);
            if (blending != null) map.put("blending", blending);
            if (vertexColors != null) map.put("vertexColors", vertexColors);
            if (fog != null) map.put("fog", fog);
            if (emissive != null) map.put("emissive", emissive);
            if (emissiveIntensity != null) map.put("emissiveIntensity", emissiveIntensity);
            if (metalness != null) map.put("metalness", metalness);
            if (roughness != null) map.put("roughness", roughness);
            if (envMapIntensity != null) map.put("envMapIntensity", envMapIntensity);
            if (flatShading != null) map.put("flatShading", flatShading);
            if (shininess != null) map.put("shininess", shininess);
            if (specular != null) map.put("specular", specular);
            if (clearcoat != null) map.put("clearcoat", clearcoat);
            if (clearcoatRoughness != null) map.put("clearcoatRoughness", clearcoatRoughness);
            if (transmission != null) map.put("transmission", transmission);
            if (ior != null) map.put("ior", ior);
            if (thickness != null) map.put("thickness", thickness);
            if (sheen != null) map.put("sheen", sheen);
            if (sheenRoughness != null) map.put("sheenRoughness", sheenRoughness);
            if (sheenColor != null) map.put("sheenColor", sheenColor);
            if (attenuationColor != null) map.put("attenuationColor", attenuationColor);
            if (attenuationDistance != null) map.put("attenuationDistance", attenuationDistance);
            if (iridescence != null) map.put("iridescence", iridescence);
            if (iridescenceIOR != null) map.put("iridescenceIOR", iridescenceIOR);
            if (iridescenceThicknessRange != null) map.put("iridescenceThicknessRange", iridescenceThicknessRange);
            if (specularIntensity != null) map.put("specularIntensity", specularIntensity);
            if (specularColor != null) map.put("specularColor", specularColor);
            if (reflectivity != null) map.put("reflectivity", reflectivity);
            if (dispersion != null) map.put("dispersion", dispersion);
            if (anisotropy != null) map.put("anisotropy", anisotropy);
            if (anisotropyRotation != null) map.put("anisotropyRotation", anisotropyRotation);
            if (mapUrl != null) map.put("mapUrl", mapUrl);
            if (normalMapUrl != null) map.put("normalMapUrl", normalMapUrl);
            if (roughnessMapUrl != null) map.put("roughnessMapUrl", roughnessMapUrl);
            if (metalnessMapUrl != null) map.put("metalnessMapUrl", metalnessMapUrl);
            if (emissiveMapUrl != null) map.put("emissiveMapUrl", emissiveMapUrl);
            if (aoMapUrl != null) map.put("aoMapUrl", aoMapUrl);
            if (alphaMapUrl != null) map.put("alphaMapUrl", alphaMapUrl);
            if (bumpMapUrl != null) map.put("bumpMapUrl", bumpMapUrl);
            if (displacementMapUrl != null) map.put("displacementMapUrl", displacementMapUrl);
            if (normalScale != null) map.put("normalScale", normalScale);
            if (bumpScale != null) map.put("bumpScale", bumpScale);
            if (displacementScale != null) map.put("displacementScale", displacementScale);
            if (displacementBias != null) map.put("displacementBias", displacementBias);
            if (aoMapIntensity != null) map.put("aoMapIntensity", aoMapIntensity);
            if (mapRepeat != null) map.put("mapRepeat", mapRepeat);
            if (mapOffset != null) map.put("mapOffset", mapOffset);
            if (mapRotation != null) map.put("mapRotation", mapRotation);
            return map;
        }
    }

    public static class ViewMaterialChangeCommandSimple {
        public String color = null;
        public Double opacity = null;
        public Boolean transparent = null;
        public Boolean visible = null;
        public Boolean wireframe = null;
        public Integer side = null;
        public Boolean fog = null;
        public String emissive = null;
        public Double emissiveIntensity = null;

        public Map<String, Object> toMap() {
            Map<String, Object> map = new HashMap<>();
            if (color != null) map.put("color", color);
            if (opacity != null) map.put("opacity", opacity);
            if (transparent != null) map.put("transparent", transparent);
            if (visible != null) map.put("visible", visible);
            if (wireframe != null) map.put("wireframe", wireframe);
            if (side != null) map.put("side", side);
            if (fog != null) map.put("fog", fog);
            if (emissive != null) map.put("emissive", emissive);
            if (emissiveIntensity != null) map.put("emissiveIntensity", emissiveIntensity);
            return map;
        }
    }
}

File: src/kt/PhantasyIslandKotlinRemoteControl/PhantasyIslandKotlinRemoteControl/radio/TypeDefView.kt
package PhantasyIslandKotlinRemoteControl.radio

import org.json.JSONObject
import org.json.JSONArray

/**
 * 修改 mesh 的可视化材质的指令。
 */
data class ViewMaterialChangeCommand(
    var materialType: String? = null,
    var color: String? = null,
    var opacity: Double? = null,
    var transparent: Boolean? = null,
    var visible: Boolean? = null,
    var wireframe: Boolean? = null,
    var side: Int? = null,
    var depthTest: Boolean? = null,
    var depthWrite: Boolean? = null,
    var alphaTest: Double? = null,
    var blending: Int? = null,
    var vertexColors: Boolean? = null,
    var fog: Boolean? = null,
    var emissive: String? = null,
    var emissiveIntensity: Double? = null,
    var metalness: Double? = null,
    var roughness: Double? = null,
    var envMapIntensity: Double? = null,
    var flatShading: Boolean? = null,
    var shininess: Double? = null,
    var specular: String? = null,
    var clearcoat: Double? = null,
    var clearcoatRoughness: Double? = null,
    var transmission: Double? = null,
    var ior: Double? = null,
    var thickness: Double? = null,
    var sheen: Double? = null,
    var sheenRoughness: Double? = null,
    var sheenColor: String? = null,
    var attenuationColor: String? = null,
    var attenuationDistance: Double? = null,
    var iridescence: Double? = null,
    var iridescenceIOR: Double? = null,
    var iridescenceThicknessRange: Pair<Double, Double>? = null,
    var specularIntensity: Double? = null,
    var specularColor: String? = null,
    var reflectivity: Double? = null,
    var dispersion: Double? = null,
    var anisotropy: Double? = null,
    var anisotropyRotation: Double? = null,
    var mapUrl: String? = null,
    var normalMapUrl: String? = null,
    var roughnessMapUrl: String? = null,
    var metalnessMapUrl: String? = null,
    var emissiveMapUrl: String? = null,
    var aoMapUrl: String? = null,
    var alphaMapUrl: String? = null,
    var bumpMapUrl: String? = null,
    var displacementMapUrl: String? = null,
    var normalScale: Pair<Double, Double>? = null,
    var bumpScale: Double? = null,
    var displacementScale: Double? = null,
    var displacementBias: Double? = null,
    var aoMapIntensity: Double? = null,
    var mapRepeat: Pair<Double, Double>? = null,
    var mapOffset: Pair<Double, Double>? = null,
    var mapRotation: Double? = null
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        materialType?.let { json.put("materialType", it) }
        color?.let { json.put("color", it) }
        opacity?.let { json.put("opacity", it) }
        transparent?.let { json.put("transparent", it) }
        visible?.let { json.put("visible", it) }
        wireframe?.let { json.put("wireframe", it) }
        side?.let { json.put("side", it) }
        depthTest?.let { json.put("depthTest", it) }
        depthWrite?.let { json.put("depthWrite", it) }
        alphaTest?.let { json.put("alphaTest", it) }
        blending?.let { json.put("blending", it) }
        vertexColors?.let { json.put("vertexColors", it) }
        fog?.let { json.put("fog", it) }
        emissive?.let { json.put("emissive", it) }
        emissiveIntensity?.let { json.put("emissiveIntensity", it) }
        metalness?.let { json.put("metalness", it) }
        roughness?.let { json.put("roughness", it) }
        envMapIntensity?.let { json.put("envMapIntensity", it) }
        flatShading?.let { json.put("flatShading", it) }
        shininess?.let { json.put("shininess", it) }
        specular?.let { json.put("specular", it) }
        clearcoat?.let { json.put("clearcoat", it) }
        clearcoatRoughness?.let { json.put("clearcoatRoughness", it) }
        transmission?.let { json.put("transmission", it) }
        ior?.let { json.put("ior", it) }
        thickness?.let { json.put("thickness", it) }
        sheen?.let { json.put("sheen", it) }
        sheenRoughness?.let { json.put("sheenRoughness", it) }
        sheenColor?.let { json.put("sheenColor", it) }
        attenuationColor?.let { json.put("attenuationColor", it) }
        attenuationDistance?.let { json.put("attenuationDistance", it) }
        iridescence?.let { json.put("iridescence", it) }
        iridescenceIOR?.let { json.put("iridescenceIOR", it) }
        iridescenceThicknessRange?.let { json.put("iridescenceThicknessRange", JSONArray(listOf(it.first, it.second))) }
        specularIntensity?.let { json.put("specularIntensity", it) }
        specularColor?.let { json.put("specularColor", it) }
        reflectivity?.let { json.put("reflectivity", it) }
        dispersion?.let { json.put("dispersion", it) }
        anisotropy?.let { json.put("anisotropy", it) }
        anisotropyRotation?.let { json.put("anisotropyRotation", it) }
        mapUrl?.let { json.put("mapUrl", it) }
        normalMapUrl?.let { json.put("normalMapUrl", it) }
        roughnessMapUrl?.let { json.put("roughnessMapUrl", it) }
        metalnessMapUrl?.let { json.put("metalnessMapUrl", it) }
        emissiveMapUrl?.let { json.put("emissiveMapUrl", it) }
        aoMapUrl?.let { json.put("aoMapUrl", it) }
        alphaMapUrl?.let { json.put("alphaMapUrl", it) }
        bumpMapUrl?.let { json.put("bumpMapUrl", it) }
        displacementMapUrl?.let { json.put("displacementMapUrl", it) }
        normalScale?.let { json.put("normalScale", JSONArray(listOf(it.first, it.second))) }
        bumpScale?.let { json.put("bumpScale", it) }
        displacementScale?.let { json.put("displacementScale", it) }
        displacementBias?.let { json.put("displacementBias", it) }
        aoMapIntensity?.let { json.put("aoMapIntensity", it) }
        mapRepeat?.let { json.put("mapRepeat", JSONArray(listOf(it.first, it.second))) }
        mapOffset?.let { json.put("mapOffset", JSONArray(listOf(it.first, it.second))) }
        mapRotation?.let { json.put("mapRotation", it) }
        return json
    }
}

/**
 * 修改 mesh 的可视化材质的指令（简化版）。
 */
data class ViewMaterialChangeCommandSimple(
    var color: String? = null,
    var opacity: Double? = null,
    var transparent: Boolean? = null,
    var visible: Boolean? = null,
    var wireframe: Boolean? = null,
    var side: Int? = null,
    var fog: Boolean? = null,
    var emissive: String? = null,
    var emissiveIntensity: Double? = null
) {
    fun toJSONObject(): JSONObject {
        val json = JSONObject()
        color?.let { json.put("color", it) }
        opacity?.let { json.put("opacity", it) }
        transparent?.let { json.put("transparent", it) }
        visible?.let { json.put("visible", it) }
        wireframe?.let { json.put("wireframe", it) }
        side?.let { json.put("side", it) }
        fog?.let { json.put("fog", it) }
        emissive?.let { json.put("emissive", it) }
        emissiveIntensity?.let { json.put("emissiveIntensity", it) }
        return json
    }
}
